### PR TITLE
MWPW-137012 - Preflight Accessibility Tab

### PR DESCRIPTION
**Pull request by Ruchika4**
* Introduce Preflight Accessibility Tab
* Performs a check to show all images missing alt-text.

Resolves: [MWPW-137012](https://jira.corp.adobe.com/browse/MWPW-137012)

**Test URLs:**
Bad Page with alt missing
- Before: https://main--milo--adobecom.hlx.page/drafts/ruchika/badpage-image?martech=off
- After: https://preflight-accessibility--milo--ruchika4.hlx.page/drafts/ruchika/badpage-image?martech=off
Screenshot:
Screen Shot 2023-09-29 at 12 50 14 PM

Good Page with alt defined
- Before: https://main--milo--adobecom.hlx.page/drafts/ruchika/goodpage-image?martech=off
- After: https://preflight-accessibility--milo--ruchika4.hlx.page/drafts/ruchika/goodpage-image?martech=off
Screenshot:
Screen Shot 2023-09-29 at 12 52 21 PM

CC:
- Before: https://main--cc--adobecom.hlx.live/creativecloud/tools/photo-editing-software?milolibs=main--milo--adobecom&martech=off
- After: https://main--cc--adobecom.hlx.live/creativecloud/tools/photo-editing-software?milolibs=preflight-accessibility--milo--ruchika4&martech=off
Screenshot:
Screen Shot 2023-09-29 at 12 53 03 PM

### CODE REVIEW

**Comment by Ruchika4** [comment]
Did some changes to design/code. Thanks for the suggestions @auniverseaway
I have updated the screenshot in the PR description. Please review and share your feedback. I still have some doubts about  the design when there are lot of images (CC screenshot) and similar images (See photoshop images in the CC screenshot).

**Review by chrischrischris — APPROVED** [review]
Assuming that more a11y checks will be added.  If not then we should probably move this into the SEO tab.

File: libs/blocks/preflight/panels/accessibility.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  if (!imagesWithoutAlt.length) {
+    result.icon = pass;
+    result.description = 'Reason: All Image are valid';
+  } else {
+    result.icon = fail;
+    result.description = 'Reason: Alt attribute or values are missing for below images on the page';
+  }
+  content.value = imagesWithoutAlt;
+  altResult.value = result;
+  return result.icon;
```
Comment (narcis-radu) [inline libs/blocks/preflight/panels/accessibility.js:31]: How is this used?

File: libs/blocks/preflight/panels/accessibility.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    if (!alt || alt.trim() === '') {
+      imagesWithoutAlt.push(img.getAttribute('src'));
+    }
+  });
+  if (!imagesWithoutAlt.length) {
+    result.icon = pass;
+    result.description = 'Reason: All Image are valid';
+  } else {
+    result.icon = fail;
+    result.description = 'Reason: Alt attribute or values are missing for below images on the page';
```
Comment (narcis-radu) [inline libs/blocks/preflight/panels/accessibility.js:27]: I think we should use constants and move them to the start of the file.
`const DESC_FAIL = 'Reason: Alt attribute or values are missing for below images on the page';`

File: libs/blocks/preflight/panels/accessibility.js (hunk 1 above)
Comment (Ruchika4) [reply libs/blocks/preflight/panels/accessibility.js]: Thanks for the review @narcis-radu It was a mistake at my side we don't need to return anything here. I have updated the code.
For **content.value** it contains all the images that are missing alt attribute or value to create the grid on modal.
**altResult.value** contains the details of message and icon to show based on checks have passed or failed

File: libs/blocks/preflight/panels/accessibility.js (hunk 2 above)
Comment (Ruchika4) [reply libs/blocks/preflight/panels/accessibility.js:27]: Haven't kept it at the start of the file as of now, as we might add more checks for accessibility in future hence keeping it at function level.

**Review by overmyheadandbody — COMMENTED** [review]
As you mentioned for the CC page, there might be instances where it's not so clear what image resource is actually referenced. Some brainstorming ideas:
* add a red/obvious/dashed outline to all images lacking an `alt` value on the page itself, visible after the modal is closed. A message should accompany this functionality;
* add unique IDs to the images lacking `alt` attributes on the page. When such an an image is clicked in the Preflight popup, close the modal and scroll to the resource that has just been clicked;
* a combination of both ideas, click problematic image in Preflight, modal is closed, scroll to the image and also add an outline to it.

I'd probably lean towards the second option, to not confuse authors with any borders, but thought it was worth mentioning.

**Comment by Ruchika4** [comment]
> As you mentioned for the CC page, there might be instances where it's not so clear what image resource is actually referenced. Some brainstorming ideas:
> 
> * add a red/obvious/dashed outline to all images lacking an `alt` value on the page itself, visible after the modal is closed. A message should accompany this functionality;
> * add unique IDs to the images lacking `alt` attributes on the page. When such an an image is clicked in the Preflight popup, close the modal and scroll to the resource that has just been clicked;
> * a combination of both ideas, click problematic image in Preflight, modal is closed, scroll to the image and also add an outline to it.
> 
> I'd probably lean towards the second option, to not confuse authors with any borders, but thought it was worth mentioning.

Thanks for your suggestions @overmyheadandbody. They are really helpful. We discussed this PR in Milo Show and tell meeting this week and it was agreed upon to share the accessibility tab with authors with current capabilities added in this PR and add more improvements as per their feedback later.

**Review by VKniaz — APPROVED** [review]

**Comment by spadmasa** [comment]
Verification done in the below milo and CC urls , able to see the accessibility tab in the preflight option displayed now. Missing altlext image is seen highlighted  which indicates the author to get the alt attribute updated in the image.
Below are the few urls which are tested :
before : 
https://main--milo--adobecom.hlx.page/drafts/ruchika/goodpage-image?martech=off
image
After :
https://preflight-accessibility--milo--ruchika4.hlx.page/drafts/ruchika/badpage-image?martech=off
image

https://preflight-accessibility--milo--ruchika4.hlx.page/drafts/ruchika/goodpage-image?martech=off
image

https://adobe.sharepoint.com/:w:/r/sites/adobecom/_layouts/15/doc2.aspx?sourcedoc=%7B58349536-A344-46B3-BBF5-3044A47FB82F%7D&file=Goodpage-Image.docx&action=default&mobileredirect=true
image


In CC  
https://main--cc--adobecom.hlx.live/creativecloud/tools/photo-editing-software?milolibs=preflight-accessibility--milo--ruchika4&martech=off
image

https://main--cc--adobecom.hlx.live/creativecloud/tools/3d-animation-software?milolibs=preflight-accessibility--milo--ruchika4&martech=off
image
https://main--cc--adobecom.hlx.live/creativecloud/animation/discover?milolibs=preflight-accessibility--milo--ruchika4&martech=off
image

https://main--cc--adobecom.hlx.live/creativecloud/file-types/image/vector?milolibs=preflight-accessibility--milo--ruchika4&martech=off
image
https://main--cc--adobecom.hlx.live/creativecloud/file-types/image/raster/ecw-file?milolibs=preflight-accessibility--milo--ruchika4&martech=off
image

**Review by overmyheadandbody — APPROVED** [review]

File: libs/blocks/preflight/panels/accessibility.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  const imagesWithoutAlt = [];
+  images.forEach((img) => {
+    const alt = img.getAttribute('alt');
+    if (!alt || alt.trim() === '') {
+      imagesWithoutAlt.push(img.getAttribute('src'));
+    }
+  });
+  if (!imagesWithoutAlt.length) {
+    result.icon = pass;
+    result.description = 'Reason: All Image are valid';
```
Comment (auniverseaway) [inline libs/blocks/preflight/panels/accessibility.js:24]: Nit: image should be plural and image should not be capitalized.
```suggestion
    result.description = 'Reason: All images are valid.';
```

File: libs/blocks/preflight/panels/accessibility.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  } else {
+    result.icon = fail;
+    result.description = 'Reason: Alt attribute or values are missing for below images on the page';
```
Comment (auniverseaway) [inline libs/blocks/preflight/panels/accessibility.js:27]: Nit: Clearer vocabulary that matches what an author would see in Word.
```suggestion
    result.description = 'Reason: Alt text missing for the following images:';
```

**Review by auniverseaway — APPROVED** [review]
Small nits with the text. Otherwise, looks great!

File: libs/blocks/preflight/panels/accessibility.js (hunk 3 above)
Comment (Ruchika4) [reply libs/blocks/preflight/panels/accessibility.js]: Fixed

File: libs/blocks/preflight/panels/accessibility.js (hunk 4 above)
Comment (Ruchika4) [reply libs/blocks/preflight/panels/accessibility.js]: Fixed

**Comment by Ruchika4** [comment]
> Small nits with the text. Otherwise, looks great!

Thanks for the review @auniverseaway Incorporated the review comments.

**Pull request merged**